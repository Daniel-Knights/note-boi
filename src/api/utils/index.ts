export * from './decorators';
export * from './error';
export * from './http';
