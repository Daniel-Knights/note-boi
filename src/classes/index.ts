export * from './appError';
export * from './debounceQueue';
export * from './encryptor';
export * from './fetchBuilder';
export * from './keyStore';
export * from './storage';
export * from './unsyncedNotesManager';
