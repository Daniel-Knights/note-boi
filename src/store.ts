import { reactive } from 'vue';
import { invoke } from '@tauri-apps/api/tauri';
import { v4 as uuidv4 } from 'uuid';

import { isEmptyNote } from './utils';

export class Note {
  readonly id = uuidv4();
  body = {
    delta: '',
    text: '',
  };
  timestamp = Date.now();
}

interface State {
  notes: Note[];
  selectedNote: Note;
}

export const state = reactive<State>({
  notes: [],
  selectedNote: new Note(),
});

/** Sorts notes in descending order by timestamp. */
function sortNotes() {
  state.notes.sort((a, b) => b.timestamp - a.timestamp);
}

/** Finds note index within {@link state.notes}. */
function findNoteIndex(id: string) {
  return state.notes.findIndex((nt) => nt.id === id);
}

/** Finds note within {@link state.notes}. */
export function findNote(id: string): Note | undefined {
  return state.notes.find((nt) => nt.id === id);
}

/** Deletes {@link state.selectedNote} when both `title` and `body` are empty. */
function clearEmptyNote(): void {
  const isValidClear = state.notes.length > 1;

  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote) return;

  if (isValidClear && isEmptyNote(foundNote)) {
    deleteNote(state.selectedNote.id);
  }
}

const selectEvent = new CustomEvent('note-select');

/**
 * Looks for note with given `id` in {@link state.notes},
 * and sets it to {@link state.selectedNote}.
 */
export function selectNote(id: string): void {
  if (state.selectedNote.id === id) return;

  clearEmptyNote();

  const foundNote = findNote(id);
  if (foundNote) state.selectedNote = { ...foundNote };

  document.dispatchEvent(selectEvent);
}

/** Fetches all notes and updates {@link state}. */
export async function getAllNotes(): Promise<void> {
  const fetchedNotes = await invoke<Note[]>('get_all_notes').catch(console.error);

  const hasNotes = fetchedNotes && fetchedNotes.length > 0;
  if (!hasNotes) return newNote();

  state.notes = fetchedNotes;

  sortNotes();

  if (isEmptyNote(fetchedNotes[0])) {
    state.notes[0].timestamp = Date.now();
  }

  state.selectedNote = { ...state.notes[0] };

  document.dispatchEvent(selectEvent);
}

/** Deletes note with the given `id`. */
export function deleteNote(id: string): void {
  state.notes.splice(findNoteIndex(id), 1);
  state.selectedNote = { ...state.notes[0] };

  sortNotes();

  invoke('delete_note', { id }).catch(console.error);
}

/** Creates an empty note. */
export function newNote(): void {
  const foundNote = findNote(state.selectedNote.id);

  // Only update timestamp if selected note is empty
  if (foundNote && isEmptyNote(foundNote)) {
    state.selectedNote.timestamp = Date.now();
    return;
  }

  const freshNote = new Note();

  state.notes.unshift(freshNote);
  state.selectedNote = { ...freshNote };

  invoke('new_note', { note: { ...freshNote } }).catch(console.error);

  document.dispatchEvent(selectEvent);
}

/** Edits note body on Quill `text-change`. */
export function editBody(delta: string, text: string): void {
  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote || text === foundNote.body.text) return;

  const timestamp = Date.now();
  foundNote.timestamp = timestamp;
  state.selectedNote.timestamp = timestamp;

  foundNote.body = { delta, text };

  sortNotes();

  invoke('edit_note', { note: { ...foundNote } }).catch(console.error);
}
