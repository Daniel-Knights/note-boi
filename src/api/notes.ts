import {
  AppError,
  DebounceQueue,
  Encryptor,
  ERROR_CODE,
  ErrorConfig,
  FetchBuilder,
  KeyStore,
} from '../classes';
import { DecryptedNoteDiff, NOTE_EVENTS } from '../constant';
import {
  findNote,
  newNote,
  noteState,
  selectNote,
  sortStateNotes,
  UnsyncedEventDetail,
} from '../store/note';
import { resetAppError, syncState } from '../store/sync';
import { isEmptyNote, tauriEmit, tauriInvoke } from '../utils';

import {
  parseErrorRes,
  resIsOk,
  route,
  throwEncryptorError,
  throwFetchError,
} from './utils';

const syncQueue = new DebounceQueue();

// Sync
export const sync = route(async (isCancelled?: () => boolean) => {
  const errorConfig = {
    code: ERROR_CODE.SYNC,
    retry: { fn: sync, args: [isCancelled] },
    display: { sync: true },
  } satisfies Omit<ErrorConfig<typeof sync>, 'message'>;

  const passwordKey = await KeyStore.getKey();
  if (!passwordKey || isCancelled?.()) return;

  const notesToEncrypt = noteState.notes.filter((nt) => {
    const noteIsEdited = syncState.unsyncedNoteIds.edited.has(nt.id);
    const noteIsCached = syncState.encryptedNotesCache.has(nt.id);

    return (noteIsEdited || !noteIsCached) && !isEmptyNote(nt);
  });

  const [accessToken, encryptedNotes] = await Promise.all([
    tauriInvoke('get_access_token', {
      username: syncState.username,
    }),
    Encryptor.encryptNotes(notesToEncrypt, passwordKey).catch((err) =>
      throwEncryptorError(errorConfig, err)
    ),
  ]);
  if (!accessToken || !encryptedNotes || isCancelled?.()) return;

  encryptedNotes.forEach((nt) => {
    syncState.encryptedNotesCache.set(nt.id, nt);
  });

  const res = await new FetchBuilder('/notes/sync')
    .method('PUT')
    .withAuth(syncState.username, accessToken)
    .body({
      notes: [...syncState.encryptedNotesCache.values()],
      deleted_note_ids: [...syncState.unsyncedNoteIds.deleted],
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res || isCancelled?.()) return;

  if (resIsOk(res)) {
    resetAppError();
    tauriEmit('auth', { is_logged_in: true });

    // Users' session must still be valid
    syncState.isLoggedIn = true;

    const decryptedNotes = await Promise.all([
      Encryptor.decryptNotes(res.data.note_diff.added, passwordKey),
      Encryptor.decryptNotes(res.data.note_diff.edited, passwordKey),
    ]).catch((err) => throwEncryptorError(errorConfig, err));
    if (!decryptedNotes || isCancelled?.()) return;

    await updateLocalNoteStateFromDiff({
      added: decryptedNotes[0],
      edited: decryptedNotes[1],
      deleted_ids: res.data.note_diff.deleted_ids,
    });
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

/**
 * Updates local note state based on the given diff.
 */
export function updateLocalNoteStateFromDiff(noteDiff: DecryptedNoteDiff) {
  let selectedNoteIsDeleted = false;
  let remoteSelectedNote;

  // Resolve edited and deleted note conflicts
  for (let i = noteState.notes.length - 1; i >= 0; i -= 1) {
    const ln = noteState.notes[i]!;
    const isSelectedNote = ln.id === noteState.selectedNote.id;

    // Remove notes that were deleted on the server
    if (noteDiff.deleted_ids.includes(ln.id)) {
      noteState.notes.splice(i, 1);

      if (isSelectedNote) {
        selectedNoteIsDeleted = true;
      }

      continue;
    }

    // Update existing notes
    const foundRemoteNoteIndex = noteDiff.edited.findIndex((rn) => rn.id === ln.id);

    if (foundRemoteNoteIndex > -1) {
      noteState.notes[i] = noteDiff.edited[foundRemoteNoteIndex]!;

      noteDiff.edited.splice(foundRemoteNoteIndex, 1);

      if (isSelectedNote) {
        remoteSelectedNote = noteState.notes[i];
      }
    }
  }

  // Add new notes from the server
  noteDiff.added.forEach((rn) => {
    noteState.notes.push(rn);
  });

  sortStateNotes();

  // New note if no notes exist
  if (noteState.notes.length === 0) {
    newNote();
  }
  // Select next note if current selected note was deleted
  else if (selectedNoteIsDeleted) {
    selectNote(noteState.notes[0]!.id);
  }
  // Update selected note if edited and ensure editor updates
  else if (remoteSelectedNote) {
    noteState.selectedNote.content = remoteSelectedNote.content;
    noteState.selectedNote.timestamp = remoteSelectedNote.timestamp;

    document.dispatchEvent(new Event(NOTE_EVENTS.change));
  }
  // Select next note if current selected note is empty and not deliberately created
  else if (noteState.notes.length > 1 && !syncState.unsyncedNoteIds.new) {
    const foundNote = findNote(noteState.selectedNote.id);

    if (!foundNote || isEmptyNote(foundNote)) {
      selectNote(noteState.notes[1]!.id);
    }
  }

  syncState.unsyncedNoteIds.clear();

  return tauriInvoke('sync_local_notes', { notes: noteState.notes });
}

/**
 * Adds sync call to the debounce queue.
 * `isInitial` indicates this is the initial sync on app load.
 */
export function debounceSync(isInitial = false): void {
  if (!isInitial && !syncState.isLoggedIn) return;

  syncQueue.add(
    (isCancelled) => {
      return sync(isCancelled);
    },
    isInitial ? undefined : 500
  );
}

// Keep track of notes with unsynced changes
document.addEventListener(
  NOTE_EVENTS.unsynced,
  (ev: CustomEventInit<UnsyncedEventDetail>) => {
    if (!ev.detail) return;

    syncState.unsyncedNoteIds.set({ [ev.detail.kind]: [ev.detail.noteId] });
  }
);
