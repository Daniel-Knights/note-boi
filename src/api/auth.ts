import {
  AppError,
  Encryptor,
  ERROR_CODE,
  ErrorConfig,
  FetchBuilder,
  KeyStore,
  Storage,
} from '../classes';
import { noteState } from '../store/note';
import { resetAppError, syncState } from '../store/sync';
import { isEmptyNote, tauriEmit, tauriInvoke } from '../utils';

import { updateLocalNoteStateFromDiff } from './notes';
import {
  parseErrorRes,
  resIsOk,
  route,
  throwEncryptorError,
  throwFetchError,
} from './utils';

export function clientSideLogout(): Promise<void> {
  tauriInvoke('delete_access_token', { username: syncState.username });

  syncState.username = '';
  syncState.isLoggedIn = false;

  Storage.remove('USERNAME');
  KeyStore.reset();

  return tauriEmit('auth', { is_logged_in: false });
}

// Login
export const login = route(async () => {
  const errorConfig = {
    code: ERROR_CODE.LOGIN,
    retry: { fn: login },
    display: {
      form: true,
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof login>, 'message'>;

  const passwordKey = await Encryptor.generatePasswordKey(syncState.password);

  const encryptedNotes = await Encryptor.encryptNotes(
    noteState.notes.filter((nt) => !isEmptyNote(nt)),
    passwordKey
  ).catch((err) => throwEncryptorError(errorConfig, err));
  if (!encryptedNotes) return;

  encryptedNotes.forEach((nt) => {
    syncState.encryptedNotesCache.set(nt.id, nt);
  });

  const res = await new FetchBuilder('/auth/login')
    .method('POST')
    .body({
      username: syncState.username,
      password: syncState.password,
      notes: encryptedNotes,
      deleted_notes: syncState.unsyncedNotes.deleted,
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    syncState.password = '';
    syncState.isLoggedIn = true;

    Storage.set('USERNAME', syncState.username);

    resetAppError();
    tauriEmit('auth', { is_logged_in: true });

    await KeyStore.storeKey(passwordKey);

    const decryptedNotes = await Promise.all([
      Encryptor.decryptNotes(res.data.note_diff.added, passwordKey),
      Encryptor.decryptNotes(res.data.note_diff.edited, passwordKey),
    ]).catch((err) => throwEncryptorError(errorConfig, err));
    if (!decryptedNotes) return;

    await updateLocalNoteStateFromDiff({
      added: decryptedNotes[0],
      edited: decryptedNotes[1],
      deleted: res.data.note_diff.deleted,
    });
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

// Signup
export const signup = route(async () => {
  const errorConfig = {
    code: ERROR_CODE.SIGNUP,
    retry: { fn: signup },
    display: {
      form: true,
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof signup>, 'message'>;

  const passwordKey = await Encryptor.generatePasswordKey(syncState.password);

  await KeyStore.storeKey(passwordKey);

  const encryptedNotes = await Encryptor.encryptNotes(
    noteState.notes.filter((nt) => !isEmptyNote(nt)),
    passwordKey
  ).catch((err) => throwEncryptorError(errorConfig, err));
  if (!encryptedNotes) return;

  const res = await new FetchBuilder('/auth/signup')
    .method('POST')
    .body({
      username: syncState.username,
      password: syncState.password,
      notes: encryptedNotes,
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    resetAppError();
    tauriEmit('auth', { is_logged_in: true });

    syncState.password = '';
    syncState.isLoggedIn = true;
    syncState.unsyncedNotes.clear();

    Storage.set('USERNAME', syncState.username);
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

// Logout
export const logout = route(async () => {
  const errorConfig = {
    code: ERROR_CODE.LOGOUT,
    retry: { fn: logout },
    display: { form: true },
  } satisfies Omit<ErrorConfig<typeof logout>, 'message'>;

  const accessToken = await tauriInvoke('get_access_token', {
    username: syncState.username,
  });

  const fetchPromise = new FetchBuilder('/auth/logout')
    .method('POST')
    .withAuth(syncState.username, accessToken)
    .fetch()
    .catch((err) => throwFetchError(errorConfig, err));

  const [res] = await Promise.all([fetchPromise, clientSideLogout()]);
  if (!res) return;

  if (resIsOk(res)) {
    resetAppError();
  } else {
    // No use bringing this to the user's attention
    throw new AppError({
      code: ERROR_CODE.LOGOUT,
      message: parseErrorRes(res),
    });
  }
});
