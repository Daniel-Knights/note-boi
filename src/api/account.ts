import * as dialog from '@tauri-apps/plugin-dialog';

import {
  AppError,
  Encryptor,
  ERROR_CODE,
  ErrorConfig,
  FetchBuilder,
  KeyStore,
} from '../classes';
import { noteState } from '../store/note';
import { resetAppError, syncState } from '../store/sync';
import { tauriInvoke } from '../utils';

import { clientSideLogout } from './auth';
import {
  parseErrorRes,
  resIsOk,
  route,
  throwEncryptorError,
  throwFetchError,
} from './utils';

export const changePassword = route(async (): Promise<void> => {
  const errorConfig = {
    code: ERROR_CODE.CHANGE_PASSWORD,
    retry: { fn: changePassword },
    display: {
      form: true,
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof changePassword>, 'message'>;

  const newKey = await Encryptor.generatePasswordKey(syncState.newPassword);

  await KeyStore.storeKey(newKey);

  const [accessToken, encryptedNotes] = await Promise.all([
    tauriInvoke('get_access_token', {
      username: syncState.username,
    }),
    Encryptor.encryptNotes(noteState.notes, newKey).catch((err) =>
      throwEncryptorError(errorConfig, err)
    ),
  ]);
  if (!accessToken || !encryptedNotes) return;

  const res = await new FetchBuilder('/account/change-password')
    .method('PUT')
    .withAuth(syncState.username, accessToken)
    .body({
      current_password: syncState.password,
      new_password: syncState.newPassword,
      notes: encryptedNotes,
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    resetAppError();

    syncState.password = '';
    syncState.newPassword = '';
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

export const deleteAccount = route(async (): Promise<void> => {
  const askRes = await dialog.ask('Are you sure?', {
    title: 'Delete account',
    kind: 'warning',
  });
  if (!askRes) return;

  const errorConfig = {
    code: ERROR_CODE.DELETE_ACCOUNT,
    retry: { fn: deleteAccount },
    display: {
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof deleteAccount>, 'message'>;

  const accessToken = await tauriInvoke('get_access_token', {
    username: syncState.username,
  });

  const res = await new FetchBuilder('/account/delete')
    .method('DELETE')
    .withAuth(syncState.username, accessToken)
    .fetch()
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    await clientSideLogout();
    resetAppError();
    syncState.unsyncedNotes.clear(true);
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});
