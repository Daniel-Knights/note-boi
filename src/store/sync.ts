import { reactive } from 'vue';
import { http, invoke } from '@tauri-apps/api';
import type { Response } from '@tauri-apps/api/http';

import { getAllNotes, Note, state as noteState } from './note';
import { isDev, tauriEmit } from '../utils';

export enum ErrorType {
  None,
  Auth,
  Push,
  Pull,
  Logout,
}

interface State {
  username: string;
  password: string;
  token: string;
  hasUnsyncedNotes: boolean;
  isLoading: boolean;
  isLogin: boolean; // For switching login/signup form
  autoSyncEnabled: boolean;
  error: {
    type: ErrorType;
    message: string;
  };
}

const autoSync = localStorage.getItem('auto-sync') as 'true' | 'false' | null;

export const state = reactive<State>({
  username: localStorage.getItem('username') || '',
  password: '',
  token: localStorage.getItem('token') || '',
  hasUnsyncedNotes: false,
  isLoading: false,
  isLogin: true,
  autoSyncEnabled: autoSync !== null ? autoSync === 'true' : true,
  error: {
    type: ErrorType.None,
    message: '',
  },
});

// Utils //

/** Parses an error response and returns a formatted message. */
function parseErrorRes(res: Response<Record<string, unknown>>) {
  const unknownErrorMessage = 'Unknown error, please try again';

  if (!res.data) return unknownErrorMessage;

  return typeof res.data.error === 'string' ? res.data.error : unknownErrorMessage;
}

/** Resets {@link state.error}. */
export function resetError(): void {
  state.error = { type: ErrorType.None, message: '' };
}

/** Wrapper for {@link http.fetch}. */
function tauriFetch<T>(
  endpoint: string,
  method: 'POST' | 'PUT',
  payload?: Record<string, unknown>
): Promise<Response<T>> {
  const baseUrl = isDev() ? 'http://localhost:8000' : 'TODO';

  return http.fetch<T>(`${baseUrl}/api${endpoint}`, {
    method,
    body: {
      type: 'Json',
      payload,
    },
  });
}

function clientSideLogout() {
  state.token = '';
  state.username = '';

  localStorage.removeItem('token');
  localStorage.removeItem('username');
}

/** Sets user preference for auto-syncing to {@link state} and `localStorage` */
export function setAutoSync(enabled: boolean): void {
  state.autoSyncEnabled = enabled;
  localStorage.setItem('auto-sync', enabled ? 'true' : 'false');
}

// Routes //

export async function login(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, string | Note[]>>('/login', 'POST', {
    username: state.username,
    password: state.password,
  });

  if (res.ok) {
    resetError();
    tauriEmit('login');

    await invoke('sync_all_local_notes', { notes: res.data.notes }).catch(console.error);
    await getAllNotes().catch(console.error);

    state.token = res.data.token as string;
    state.password = '';

    localStorage.setItem('username', state.username);
    localStorage.setItem('token', state.token);
  }

  state.isLoading = false;

  if (!res.ok) {
    state.error = {
      type: ErrorType.Auth,
      message: parseErrorRes(res),
    };

    console.error(res.data);
  }
}

export async function signup(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, string>>('/signup', 'POST', {
    username: state.username,
    password: state.password,
    notes: noteState.notes,
  });

  state.isLoading = false;

  if (res.ok) {
    resetError();
    tauriEmit('login');

    state.token = res.data.token;
    state.password = '';

    localStorage.setItem('username', state.username);
    localStorage.setItem('token', state.token);
  } else {
    state.error = {
      type: ErrorType.Auth,
      message: parseErrorRes(res),
    };

    console.error(res.data);
  }
}

export async function logout(): Promise<void> {
  if (!state.token) return; // Prevent bug where event.emit triggers event.listen

  state.isLoading = true;

  const res = await tauriFetch<Record<string, never | string>>('/logout', 'POST', {
    username: state.username,
    token: state.token,
  });

  state.isLoading = false;

  if (res.ok) {
    resetError();
    tauriEmit('logout');
    clientSideLogout();
  } else {
    state.error = {
      type: ErrorType.Logout,
      message: parseErrorRes(res),
    };
  }
}

export async function pull(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, string | Note[]>>('/notes', 'POST', {
    username: state.username,
    token: state.token,
  });

  if (res.ok) {
    resetError();
    await invoke('sync_all_local_notes', { notes: res.data.notes }).catch(console.error);
    await getAllNotes().catch(console.error);
  }

  state.isLoading = false;

  if (!res.ok) {
    state.error = {
      type: ErrorType.Pull,
      message: parseErrorRes(res),
    };

    // User not found
    if (res.status === 404) clientSideLogout();

    console.error(res.data);
  }
}

export async function push(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, never | string>>('/notes', 'PUT', {
    username: state.username,
    token: state.token,
    notes: noteState.notes,
  });

  state.isLoading = false;

  if (res.ok) {
    resetError();
    state.hasUnsyncedNotes = false;
  } else {
    state.error = {
      type: ErrorType.Push,
      message: parseErrorRes(res),
    };

    console.error(res.data);
  }
}

document.addEventListener('note-unsynced', () => {
  if (state.token) {
    state.hasUnsyncedNotes = true;
  }
});
