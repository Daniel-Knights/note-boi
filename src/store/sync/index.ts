import { http } from '@tauri-apps/api';
import type { Response } from '@tauri-apps/api/http';
import { reactive } from 'vue';

import { STORAGE_KEYS } from '../../constant';
import { isDev } from '../../utils';
import { UnsyncedNoteIds, unsyncedNoteIds } from './note';

export enum ErrorType {
  None,
  Auth,
  Push,
  Pull,
  Logout,
}

export const syncState = reactive({
  username: localStorage.getItem(STORAGE_KEYS.USERNAME) || '',
  password: '',
  token: localStorage.getItem(STORAGE_KEYS.TOKEN) || '',
  isLoading: false,
  isLogin: true, // For switching login/signup form
  error: {
    type: ErrorType.None,
    message: '',
  },
  unsyncedNoteIds: <UnsyncedNoteIds>{
    new: unsyncedNoteIds.new || '',
    edited: new Set<string>(unsyncedNoteIds.edited),
    deleted: new Set<string>(unsyncedNoteIds.deleted),
    get size() {
      return this.edited.size + this.deleted.size;
    },
    clear() {
      this.new = '';
      this.edited.clear();
      this.deleted.clear();
      localStorage.removeItem(STORAGE_KEYS.UNSYNCED);
    },
    add(ids): void {
      if (ids.new !== undefined) this.new = ids.new;

      ids.edited?.forEach((id) => this.edited.add(id));

      // Only track deleted notes if logged in
      ids.deleted?.forEach((id) => {
        this.deleted.add(id);

        if (this.edited.has(id)) {
          this.edited.delete(id);
        }
      });

      if (this.edited.has(this.new) || this.deleted.has(this.new)) {
        this.new = '';
      }

      localStorage.setItem(
        STORAGE_KEYS.UNSYNCED,
        JSON.stringify({
          new: this.new,
          edited: [...this.edited],
          deleted: [...this.deleted],
        })
      );
    },
  },
});

// Utils //

/** Parses an error response and returns a formatted message. */
export function parseErrorRes(res: Response<Record<string, unknown>>): string {
  const unknownErrorMessage = 'Unknown error, please try again';

  if (!res.data) return unknownErrorMessage;

  return typeof res.data.error === 'string' ? res.data.error : unknownErrorMessage;
}

/** Resets {@link state.error}. */
export function resetError(): void {
  syncState.error = { type: ErrorType.None, message: '' };
}

/** Wrapper for {@link http.fetch}. */
export function tauriFetch<T>(
  endpoint: string,
  method: 'POST' | 'PUT',
  payload?: Record<string, unknown>
): Promise<Response<T>> {
  const baseUrl = isDev()
    ? 'http://localhost:8000'
    : 'https://note-boi-server.herokuapp.com';

  return http.fetch<T>(`${baseUrl}/api${endpoint}`, {
    method,
    body: {
      type: 'Json',
      payload,
    },
  });
}

/** Catches hanging requests (e.g. due to server error). */
export function catchHang(err: unknown, type: ErrorType): void {
  console.error(err);

  syncState.isLoading = false;
  syncState.error = {
    type,
    message: 'Request failed',
  };
}

export { autoPush, push, pull, syncNotes, UnsyncedNoteIds } from './note';
export { clientSideLogout, login, signup, logout } from './auth';
export { deleteAccount } from './account';
