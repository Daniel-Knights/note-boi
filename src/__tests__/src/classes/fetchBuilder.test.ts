import * as s from '../../../store/sync';
import { FetchBuilder } from '../../../classes';
import { mockApi, mockKeyring } from '../../api';

describe('FetchBuilder', () => {
  const testHeaders = {
    'X-Test-Header': 'test',
  };

  it('Without auth', async () => {
    const { calls } = mockApi();

    const fetchSpy = vi.spyOn(window, 'fetch');
    const body = {
      username: 'd',
      password: '1',
    };

    s.syncState.loadingCount = 1; // Mock request expects this

    const res = await new FetchBuilder('/login')
      .method('POST')
      .headers(testHeaders)
      .body(body)
      .fetch('d');

    expect(fetchSpy).toHaveBeenCalledOnce();
    expect(fetchSpy).toHaveBeenCalledWith(`${FetchBuilder.serverUrl}/api/login`, {
      method: 'POST',
      body: JSON.stringify(body),
      headers: {
        ...FetchBuilder.defaultHeaders,
        ...testHeaders,
      },
    });

    assert.strictEqual(calls.size, 2);
    assert.isTrue(calls.invoke.has('set_access_token'));
    assert.deepEqual(calls.invoke[0]!.calledWith, {
      username: 'd',
      accessToken: 'test-token',
    });
    assert.isTrue(calls.request.has('/login'));
    assert.deepEqual(res, {
      ok: true,
      data: {
        notes: [],
        access_token: 'test-token',
      },
      status: 200,
    });
  });

  it('With auth', async () => {
    const { calls } = mockApi();

    const fetchSpy = vi.spyOn(window, 'fetch');

    // Mock request expects these
    s.syncState.loadingCount = 1;
    mockKeyring.d = 'test-token';

    const res = await new FetchBuilder('/notes/pull')
      .method('POST')
      .headers(testHeaders)
      .withAuth('d', 'test-token')
      .fetch('d');

    expect(fetchSpy).toHaveBeenCalledOnce();
    expect(fetchSpy).toHaveBeenCalledWith(`${FetchBuilder.serverUrl}/api/notes/pull`, {
      method: 'POST',
      credentials: 'same-origin',
      headers: {
        ...FetchBuilder.defaultHeaders,
        ...testHeaders,
        'X-Username': 'd',
        Authorization: 'Bearer test-token',
      },
    });

    assert.strictEqual(calls.size, 2);
    assert.isTrue(calls.invoke.has('set_access_token'));
    assert.deepEqual(calls.invoke[0]!.calledWith, {
      username: 'd',
      accessToken: 'test-token',
    });
    assert.isTrue(calls.request.has('/notes/pull'));
    assert.deepEqual(res, {
      ok: true,
      data: {
        notes: [],
        access_token: 'test-token',
      },
      status: 200,
    });
  });
});
