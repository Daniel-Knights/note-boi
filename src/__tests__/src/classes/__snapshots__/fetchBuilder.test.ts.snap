// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FetchBuilder > With auth > KeyStore.getKey() 1`] = `undefined`;

exports[`FetchBuilder > With auth > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`FetchBuilder > With auth > Storage.get('THEME') 1`] = `null`;

exports[`FetchBuilder > With auth > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`FetchBuilder > With auth > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`FetchBuilder > With auth > Storage.get('USERNAME') 1`] = `null`;

exports[`FetchBuilder > With auth > allCalls 1`] = `
[
  [
    "request",
    {
      "calledWith": {
        "body": undefined,
        "headers": {
          "Authorization": "Bearer test-token",
          "Content-Security-Policy": "default-src 'self'; connect-src http://localhost:8000;",
          "Content-Type": "application/json",
          "X-Test-Header": "test",
          "X-Username": "d",
        },
        "method": "POST",
      },
      "name": "/notes/pull",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "accessToken": "test-token",
        "username": "d",
      },
      "name": "set_access_token",
    },
  ],
]
`;

exports[`FetchBuilder > With auth > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`FetchBuilder > With auth > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`FetchBuilder > With auth > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`FetchBuilder > With auth > openedPopup 1`] = `undefined`;

exports[`FetchBuilder > With auth > selectedTheme 1`] = `"System"`;

exports[`FetchBuilder > With auth > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 1,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`FetchBuilder > With auth > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`FetchBuilder > Without auth > KeyStore.getKey() 1`] = `undefined`;

exports[`FetchBuilder > Without auth > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`FetchBuilder > Without auth > Storage.get('THEME') 1`] = `null`;

exports[`FetchBuilder > Without auth > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`FetchBuilder > Without auth > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`FetchBuilder > Without auth > Storage.get('USERNAME') 1`] = `null`;

exports[`FetchBuilder > Without auth > allCalls 1`] = `
[
  [
    "request",
    {
      "calledWith": {
        "body": "{"username":"d","password":"1"}",
        "headers": {
          "Content-Security-Policy": "default-src 'self'; connect-src http://localhost:8000;",
          "Content-Type": "application/json",
          "X-Test-Header": "test",
        },
        "method": "POST",
      },
      "name": "/auth/login",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "accessToken": "test-token",
        "username": "d",
      },
      "name": "set_access_token",
    },
  ],
]
`;

exports[`FetchBuilder > Without auth > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`FetchBuilder > Without auth > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`FetchBuilder > Without auth > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`FetchBuilder > Without auth > openedPopup 1`] = `undefined`;

exports[`FetchBuilder > Without auth > selectedTheme 1`] = `"System"`;

exports[`FetchBuilder > Without auth > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 1,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`FetchBuilder > Without auth > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
